homeassistant/components/light/xiaomi_miio.py

```diff
@@ -63,27 +85,58 @@ def async_setup_platform(hass, config, async_add_devices, discovery_info=None):
             from miio import PhilipsEyecare
             light = PhilipsEyecare(host, token)
             device = XiaomiPhilipsEyecareLamp(name, light, device_info)
-            devices.append(device)
         elif device_info.model == 'philips.light.ceiling':
             from miio import Ceil
             light = Ceil(host, token)
             device = XiaomiPhilipsCeilingLamp(name, light, device_info)
-            devices.append(device)
         elif device_info.model == 'philips.light.bulb':
             from miio import PhilipsBulb
             light = PhilipsBulb(host, token)
             device = XiaomiPhilipsLightBall(name, light, device_info)
-            devices.append(device)
         else:
             _LOGGER.error(
                 'Unsupported device found! Please create an issue at '
                 'https://github.com/rytilahti/python-miio/issues '
                 'and provide the following data: %s', device_info.model)
+            return False
 
     except DeviceException:
         raise PlatformNotReady
 
-    async_add_devices(devices, update_before_add=True)
+    hass.data[PLATFORM][host] = device
+    async_add_devices([device], update_before_add=True)
+
+    @asyncio.coroutine
+    def async_service_handler(service):
+        """Map services to methods on Xiaomi Philips Lights."""
+        method = SERVICE_TO_METHOD.get(service.service)
+        params = {key: value for key, value in service.data.items()
+                  if key != ATTR_ENTITY_ID}
+        entity_ids = service.data.get(ATTR_ENTITY_ID)
+        if entity_ids:
+            target_devices = [dev for dev in hass.data[PLATFORM].values()
+                              if dev.entity_id in entity_ids]
+        else:
+            target_devices = hass.data[PLATFORM].values()
+
+        update_tasks = []
+        for target_device in target_devices:
+            yield from getattr(target_device, method['method'])(**params)
+            update_tasks.append(target_device.async_update_ha_state(True))
+
+        if update_tasks:
+            yield from asyncio.wait(update_tasks, loop=hass.loop)
+
+    descriptions = yield from hass.async_add_job(
+        load_yaml_config_file, os.path.join(
+            os.path.dirname(__file__), 'xiaomi_miio_services.yaml'))
+
+    for xiaomi_miio_service in SERVICE_TO_METHOD:
+        schema = SERVICE_TO_METHOD[xiaomi_miio_service].get(
+            'schema', XIAOMI_MIIO_SERVICE_SCHEMA)
+        hass.services.async_register(
+            DOMAIN, xiaomi_miio_service, async_service_handler,
+            description=descriptions.get(xiaomi_miio_service), schema=schema)
```

Comment:
Services are no longer passed in during registration but are instead automatically discovered by looking at `<DOMAIN>/services.yaml`. Please move the descriptions there.